### PR TITLE
Plane: maximum altitude for FBWB

### PR DESCRIPTION
**Pull request by mmk0102**
To fix issue #17019: Plane: maximum altitude for FBWB and cruise mode.
Add ALT_FBWCM_MAX param for maximum altitude FBWB. Default (zero) - no limit.

### CODE REVIEW

File: ArduPlane/Parameters.cpp [hunk 1]
```diff
@@ -681,6 +681,13 @@ const AP_Param::Info Plane::var_info[] = {
     // @User: Standard
     GSCALAR(FBWB_min_altitude_cm,   "ALT_HOLD_FBWCM", ALT_HOLD_FBW_CM),
 
+    // @Param: ALT_FBWCM_MAX
+    // @DisplayName: Maximum altitude for FBWB and CRUISE modes
+    // @Description: This is the maximum altitude in meters that FBWB and CRUISE modes will allow. If you attempt to descend upper this altitude the plane will hold level. A value of zero means no limit.
+    // @Units: m
+    // @User: Standard
+    GSCALAR(FBWB_max_altitude_m,   "ALT_FBWCM_MAX", 0),
```
Comment (Hwurzburg) [inline ArduPlane/Parameters.cpp:689]: variables do not have caps AFAIK....   think it should be fbwb_max_alt_m

File: ArduPlane/altitude.cpp [hunk 2]
```diff
@@ -352,10 +352,19 @@ int32_t Plane::calc_altitude_error_cm(void)
...
 
 /*
-  check for FBWB_min_altitude_cm violation
+  check for min and max altitude: FBWB_min_altitude_cm and FBWB_max_altitude_m violation
  */
-void Plane::check_fbwb_minimum_altitude(void)
+void Plane::check_fbwb_altitude(void)//check_fbwb_minimum_altitude
 {
+    //check for max
+    if (g.FBWB_max_altitude_m != 0)
```
Comment (Hwurzburg) [inline ArduPlane/altitude.cpp:360]: test should be >0...often people use -1 to disable things, irrespective of what the metadata says

**Review by Hwurzburg — CHANGES_REQUESTED** [review]

File: ArduPlane/altitude.cpp (hunk 2 above)
Comment (IamPete1) [reply ArduPlane/altitude.cpp:360]: should also check its larger than alt min

File: ArduPlane/altitude.cpp [hunk 3]
```diff
@@ -352,10 +352,19 @@ int32_t Plane::calc_altitude_error_cm(void)
...
+    {
+        if (target_altitude.amsl_cm > home.alt + g.FBWB_max_altitude_m*1000) {
+            target_altitude.amsl_cm = home.alt + g.FBWB_max_altitude_m*1000;
```
Comment (IamPete1) [inline ArduPlane/altitude.cpp:363]: better to constrain, with the min check.

File: ArduPlane/Parameters.cpp (hunk 1 above)
Comment (mmk0102) [reply ArduPlane/Parameters.cpp:689]: Hi! You mean a "previous" as ### k_param_FBWB_min_altitude_cm? 
Should I correct it too ?

File: ArduPlane/altitude.cpp (hunk 2 above)
Comment (mmk0102) [reply ArduPlane/altitude.cpp:360]: mb larger than alt min and some const ?
Is 1 m correct max alt? 5 m? 10 m?
I wonna do some like:
 **if** (g.FBWB_max_altitude_m*1000 > g.FBWB_min_altitude_cm **&&** g.FBWB_max_altitude_m > **10**)

File: ArduPlane/altitude.cpp (hunk 3 above)
Comment (IamPete1) [reply ArduPlane/altitude.cpp]: You need to also consider terrain, with min alt relative to terrain and max alt above home you could end up in the situation where the pilot has no altitude control it just flys at the min terrain altitude all the time.

**Review by IamPete1 — CHANGES_REQUESTED** [review]
Need to consider terrain

**Pull request closed without merging**